pkg/chunk/cache/fifo_cache.go

```diff
@@ -179,91 +191,130 @@ func (c *FifoCache) Put(ctx context.Context, keys []string, values []interface{}
 	}
 }
 
+func (c *FifoCache) addToHead(entry *cacheEntry) {
+	if entry == nil {
+		return
+	}
+	entry.prev, entry.next = nil, c.first
+	if c.first != nil {
+		c.first.prev = entry
+	}
+	c.first = entry
+	if c.last == nil {
+		c.last = entry
+	}
+}
+
+func (c *FifoCache) deleteFromTail() *cacheEntry {
+	if c.last == nil {
+		return nil
+	}
+	ret := c.last
+	c.last = ret.prev
+	if c.last == nil {
+		c.first = nil
+	} else {
+		c.last.next = nil
+	}
+	return ret
+}
+
+func (c *FifoCache) deleteFromList(entry *cacheEntry) {
+	if entry == nil {
+		return
+	}
+	if entry.prev == nil {
+		c.first = entry.next
+	} else {
+		entry.prev.next = entry.next
+	}
+	if entry.next == nil {
+		c.last = entry.prev
+	} else {
+		entry.next.prev = entry.prev
+	}
+}
+
 func (c *FifoCache) put(ctx context.Context, key string, value interface{}) {
 	// See if we already have the entry
-	index, ok := c.index[key]
+	entry, ok := c.entries[key]
 	if ok {
-		entry := c.entries[index]
-		deltaSize := sizeOf(value) - sizeOf(entry.value)
-
-		entry.updated = time.Now()
-		entry.value = value
-
 		// Remove this entry from the FIFO linked-list.
-		c.entries[entry.prev].next = entry.next
-		c.entries[entry.next].prev = entry.prev
-
-		// Corner case: updating last element
-		if c.last == index {
-			c.last = entry.prev
+		c.deleteFromList(entry)
+		entrySz := sizeOf(entry)
+		c.currSize -= entrySz
+
+		// New entry size
+		entrySz += (sizeOfInterface(value) - sizeOfInterface(entry.value))
+
+		if c.maxSizeBytes > 0 && c.currSize+entrySz > c.maxSizeBytes {
+			// New entry does not fit in the cache. Delete the entry.
+			delete(c.entries, entry.key)
+			c.entriesCurrent.Dec()
+			c.entriesEvicted.Inc()
+			entry = nil
+		} else {
+			// Insert it at the beginning.
+			entry.updated = time.Now()
+			entry.value = value
+			c.addToHead(entry)
+			c.currSize += entrySz
 		}
+		c.memoryBytes.Set(float64(c.currSize))
+		return
+	}
 
-		// Insert it at the beginning
-		entry.next = c.first
-		entry.prev = c.last
-		c.entries[entry.next].prev = index
-		c.entries[entry.prev].next = index
-		c.first = index
+	entry = &cacheEntry{
+		updated: time.Now(),
+		key:     key,
+		value:   value,
+	}
+	entrySz := sizeOf(entry)
 
-		c.entries[index] = entry
-		c.memoryBytes.Add(float64(deltaSize))
+	// Check is the entry fits in the cache
+	if c.maxSizeBytes > 0 && entrySz > c.maxSizeBytes {
```

Comment:
This check should be moved to the top of the function and done as first thing.

Reply:
Disagree. If we already have an entry with that key, and the new entry doesn't fit the cache, we need to remove the old entry first, and then and ignore the new one.

Reply:
You're right. Resolving this comment.